File: common/windows/delphi/general/Glossary.pas
(*
  Name:             Glossary
  Copyright:        Copyright (C) SIL International.
  Documentation:    
  Description:      
  Create Date:      20 Jun 2006

  Modified Date:    28 May 2014
  Authors:          mcdurdin
  Related Files:    
  Dependencies:     

  Bugs:             
  Todo:             
  Notes:            
  History:          20 Jun 2006 - mcdurdin - Initial version
                    02 Aug 2006 - mcdurdin - Add HKLIsIME function (tests if IMM is installed before ImmIsIME)
                    23 Aug 2007 - mcdurdin - I981 - Fix ImmIsIME to be always ignored on Vista
                    03 May 2011 - mcdurdin - I2890 - Record diagnostic data when encountering registry errors
                    18 May 2012 - mcdurdin - I3306 - V9.0 - Remove TntControls + Win9x support
                    01 Jan 2013 - mcdurdin - I3724 - V9.0 - GetLanguageName function needs to be general
                    17 Jan 2013 - mcdurdin - I3763 - V9.0 - GetLanguageName refactor
                    23 Jan 2013 - mcdurdin - I3774 - V9.0 - Regression KMCOMAPI IKeymanLanguage returns region-free language name, incorrectly
                    16 Apr 2014 - mcdurdin - I4169 - V9.0 - Mnemonic layouts should be recompiled to positional based on user-selected base keyboard
                    28 May 2014 - mcdurdin - I4220 - V9.0 - Remove references to LoadKeyboardLayout, Preload, Substitutes, etc. and use only TSF
*)
unit Glossary;  // I3306

interface

uses
  System.Classes,
  System.SysUtils,
  Winapi.Imm,
  Winapi.Windows,

  ErrorControlledRegistry,
  RegistryKeys;

type
  PHKL = ^HKL;

function KeyboardIDToHKL(KeyboardID: DWord): HKL; deprecated;   // I4220
function HKLToKeyboardID(ahkl: HKL): DWord;
function HKLToLanguageID(ahkl: HKL): Word;
function HKLToLayoutNumber(ahkl: HKL): Word;
function HKLToLayoutID(ahkl: HKL): Word;
function MAKELCID(lgid, srtid: Word): DWord;
function MAKELANGID(prilangid, sublangid: Word): Word;
function IsKeymanLayout(layout: DWord): Boolean;
function IsKeymanHKL(ahkl: HKL): Boolean;

function GetLanguageName(langid: Word; var LanguageName: string; IncludeRegion: Boolean = False): Boolean;   // I3724   // I3763   // I3774

function GetKeyboardLayouts(var count: Integer): PHKL; // I4169

implementation

uses
  GetOsVersion;

function GetKeyboardLayouts(var count: Integer): PHKL;   // I4169
begin
  Result := nil;
	count := GetKeyboardLayoutList(0, Result^);
	Result := AllocMem(count * SizeOf(HKL));
  if Assigned(Result) then
  	GetKeyboardLayoutList(count, Result^);
end;

function KeyboardIDToHKL(KeyboardID: DWord): HKL;
var
  i, n: Integer;
  hp, hkls: PHKL;
begin
  Result := 0;

  hkls := nil;
	n := GetKeyboardLayoutList(0, hkls^);
	hkls := AllocMem(n * SizeOf(HKL));
  if not Assigned(hkls) then Exit;

  hp := hkls;
	GetKeyboardLayoutList(n, hkls^);
	for i := 0 to n-1 do
		if HKLToKeyboardID(hp^) = KeyboardID then
    begin
      Result := hp^;
      Break;
    end;

	FreeMem(hkls);
end;

function HKLToKeyboardID(ahkl: HKL): DWord;
var
	i, LayoutID: Integer;
  str: TStringList;
begin
	// Test for IME, and if IME return LanguageID

  //if((HIWORD(hkl) & 0xF000) == 0xE000) return (DWORD) hkl;

	// Test for standard layouts: 0409041D for instance
	// specialised layout is: F0020409 fr instance [find layout id:0002]

  if LOWORD(ahkl) = 0 then
  begin
    // Look up the HKL in the registry to get substitutes
    with TRegistryErrorControlled.Create do  // I2890
    try
      if OpenKeyReadOnly('\'+SRegKey_KeyboardLayoutSubstitutes_CU) and
        ValueExists(IntToHex(HIWORD(ahkl), 8)) then
      begin
        Result := DWord(StrToIntDef('$'+ReadString(IntToHex(HIWORD(ahkl), 8)), 0));
        Exit;
      end;
    finally
      Free;
    end;
  end;

	LayoutID := HKLToLayoutID(ahkl);
	if LayoutID = 0 then
  begin
    Result := HIWORD(ahkl);
    Exit;
  end;

	// Find the KeyboardID associated with the LayoutID

  str := TStringList.Create;
  with TRegistryErrorControlled.Create do  // I2890
  try
    RootKey := HKEY_LOCAL_MACHINE;
    if not OpenKeyReadOnly(SRegKey_KeyboardLayouts_LM) then
    begin
      Result := LOWORD(ahkl);
      Exit;
    end;

    GetKeyNames(str);
    for i := 0 to str.Count - 1 do
      if OpenKeyReadOnly('\'+SRegKey_KeyboardLayouts_LM+'\'+str[i]) and
        ValueExists(SRegValue_KeyboardLayoutID) then
      begin
        if StrToIntDef('$'+ReadString(SRegValue_KeyboardLayoutID), 0) = LayoutID then
        begin
          Result := DWord(StrToIntDef('$'+str[i], 0));
          if Result = 0 then Result := LOWORD(ahkl);
          Exit;
        end;
      end;
  finally
    Free;
    str.Free;
  end;

  Result := LOWORD(ahkl);		// should never happen
end;

function HKLToLanguageID(ahkl: HKL): Word;
begin
  Result := LOWORD(ahkl);
end;

function HKLToLayoutNumber(ahkl: HKL): Word;
var
  LayoutID: Word;
  i: Integer;
  str: TStringList;
begin
  Result := 0;

  LayoutID := HKLToLayoutID(ahkl);
  if LayoutID = 0 then Exit;

  str := TStringList.Create;
  with TRegistryErrorControlled.Create do  // I2890
  try
    RootKey := HKEY_LOCAL_MACHINE;
    if not OpenKeyReadOnly(SRegKey_KeyboardLayouts_LM) then Exit;
    GetKeyNames(str);
    for i := 0 to str.Count - 1 do
      if OpenKeyReadOnly('\'+SRegKey_KeyboardLayouts_LM+'\'+str[i]) and
        ValueExists(SRegValue_KeyboardLayoutID) then
          if StrToIntDef('$' + ReadString(SRegValue_KeyboardLayoutID), 0) = LayoutID then
        begin
          Result := HIWORD(DWord(StrToIntDef('$'+str[i], 0)));
          Exit;
        end;
  finally
    Free;
    str.Free;
  end;
end;

function HKLToLayoutID(ahkl: HKL): Word;
begin
  if (HIWORD(ahkl) and $F000) <> $F000
    then Result := 0
    else Result := HIWORD(ahkl) and $0FFF;
end;

function MAKELCID(lgid, srtid: Word): DWord;
begin
  Result := (DWord(srtid) shl 16) or lgid;
end;

function IsKeymanHKL(ahkl: HKL): Boolean;
begin
  Result := IsKeymanLayout(HKLToKeyboardID(ahkl));
end;

function IsKeymanLayout(layout: DWord): Boolean;
begin
  Result := LOWORD(layout) = $05FE;
end;

function MAKELANGID(prilangid, sublangid: Word): Word;
begin
  Result := prilangid or (sublangid shl 10);
end;


const
  LOCALE_SLANGDISPLAYNAME = $6f;

function GetLanguageName(langid: Word; var LanguageName: string; IncludeRegion: Boolean = False): Boolean;   // I3724   // I3763  // I3774
var
  szLangName: array[0..MAX_PATH] of char;
  LCType: Cardinal;
begin
  if IncludeRegion
    then LCType := LOCALE_SLANGUAGE  // I3774
    else LCType := LOCALE_SLANGDISPLAYNAME;

  Result := GetLocaleInfo(langid, LCType, szLangName, MAX_PATH) <> 0;
  if Result
    then LanguageName := Copy(szLangName, 0, MAX_PATH - 1)
    else LanguageName := 'Unknown '+IntToHex(langid, 8);
end;

end.

File: windows/src/engine/keyman32/glossary.cpp
/*
  Name:             glossary
  Copyright:        Copyright (C) SIL International.
  Documentation:
  Description:
  Create Date:      20 Jul 2008

  Modified Date:    28 May 2014
  Authors:          mcdurdin
  Related Files:
  Dependencies:

  Bugs:
  Todo:
  Notes:
  History:          20 Jul 2008 - mcdurdin - I1498 - Fix keyboard switching for Shadow keyboards on Vista+
                    20 Jul 2008 - mcdurdin - I1546 - Fix language switch with ids >= x80000000
                    20 Jul 2008 - mcdurdin - I1545 - Fix registry leak
                    11 Mar 2009 - mcdurdin - I1894 - Fix threading bugs introduced in I1888
                    11 Dec 2009 - mcdurdin - I934 - x64 - Initial version
                    12 Mar 2010 - mcdurdin - I934 - x64 - Complete
                    12 Mar 2010 - mcdurdin - I2229 - Remove hints and warnings
                    04 May 2010 - mcdurdin - I2351 - Robustness - verify _td return value
                    23 Dec 2011 - mcdurdin - I3191 - Corruption of keyboard selection cache between x86 and x64
                    04 Nov 2012 - mcdurdin - I3526 - V9.0 - Merge of I3191 - Corruption of keyboard selection cache between x86 and x64
                    28 May 2014 - mcdurdin - I4220 - V9.0 - Remove references to LoadKeyboardLayout, Preload, Substitutes, etc. and use only TSF
*/
#include "pch.h"
   // I4220
#ifndef SM_IMMENABLED
#define SM_IMMENABLED 82
#endif

#pragma warning(disable: 4996)
BOOL HKLIsIME(HKL hkl)  // I1498 - fix keyboard switching for shadow keyboards on Vista+
{
  BOOL r;
  if( (GetVersion() & 0xFF) >= 6 ) return FALSE;
  if( (GetVersion() & 0x8000000) == 0x8000000 || (GetVersion() & 0xFF) == 4 )
    r = GetSystemMetrics(SM_DBCSENABLED);
  else
    r = GetSystemMetrics(SM_IMMENABLED);

  return r && ImmIsIME(hkl);
}
#pragma warning(default: 4996)

HKL ForceKeymanIDToHKL(DWORD KeymanID)  // I3191   // I3526
{
#ifdef _WIN64
  return (HKL) (LONG_PTR) (INT64) (LONG) KeymanID;  // Sign extension for x64
#else
  return (HKL) (DWORD_PTR) KeymanID;
#endif
}

DWORD ForceHKLToKeymanID(HKL hkl)  // I3191   // I3526
{
  return (DWORD) (DWORD_PTR) hkl; // Just truncate for x64
}

DWORD HKLToKeyboardID(HKL hkl)
{
	HKEY hkey, hsubkey;
	char str[16], str2[16];
	DWORD i = 0;
	DWORD n;
	FILETIME ft;
	DWORD len;

  //SendDebugMessageFormat("hkl=%x", hkl);

	// Test for IME, and if IME return LanguageID

	if(HKLIsIME(hkl)) // I1498 - was ImmIsIME
  {
    //SendDebugMessageFormat("return HKLIsIME is true=%x", hkl);
    return ForceHKLToKeymanID(hkl);  // MCD 19/03/02  // I3191   // I3526
  }
	//if((HIWORD(hkl) & 0xF000) == 0xE000) return (DWORD) hkl;

	// Test for standard layouts: 0409041D for instance
	// specialised layout is: F0020409 fr instance [find layout id:0002]

	DWORD LayoutID = HKLToLayoutID(hkl);
	if(LayoutID == 0)
  {
    //SendDebugMessageFormat("no LayoutID, so return HIWORD(hkl)=%x", HIWORD(hkl));
    return HIWORD(hkl);
  }

 //SendDebugMessageFormat("LayoutID=%x", LayoutID);

	// Find the KeyboardID associated with the LayoutID

	if(RegOpenKeyEx(HKEY_LOCAL_MACHINE, REGSZ_SystemKeyboardLayouts, NULL, KEY_READ, &hkey) != ERROR_SUCCESS)
  {
    //SendDebugMessageFormat("fails, return LOWORD(hkl)=%x", LOWORD(hkl));
		return (DWORD) LOWORD(hkl);
  }

	for(len=16, n = i = 0; RegEnumKeyEx(hkey, i, str, &len, 0, NULL, NULL, &ft) == ERROR_SUCCESS;
		len = 16, i++, n=0)
	{
		RegOpenKeyEx(hkey, str, NULL, KEY_READ, &hsubkey);
		len = 16;
		if(RegQueryValueEx(hsubkey, REGSZ_LayoutID, NULL, NULL, (LPBYTE) str2, &len) == ERROR_SUCCESS)
		{
			n = strtoul(str2, NULL, 16);  // I1546
			if(n == LayoutID)
      {
        RegCloseKey(hsubkey); // I1545
        RegCloseKey(hkey);
      	n = strtoul(str, NULL, 16); // I1546
        SendDebugMessageFormat("return keyboard from LayoutID = %x [%s]", n, str);
        return n;
      }
		}
		RegCloseKey(hsubkey);
	}

	RegCloseKey(hkey);

  //SendDebugMessageFormat("fails[2], return LOWORD(hkl)=%x", LOWORD(hkl));
  return (DWORD) LOWORD(hkl);		// should never happen
}


WORD HKLToLanguageID(HKL hkl)
{
	return LOWORD(hkl);
}


WORD HKLToLayoutNumber(HKL hkl)
{
	WORD LayoutID = HKLToLayoutID(hkl);
	HKEY hkey, hsubkey;
	DWORD len, n, i;
	FILETIME ft;
	char str[16], str2[16];

	if(RegOpenKeyEx(HKEY_LOCAL_MACHINE, REGSZ_SystemKeyboardLayouts, NULL, KEY_READ, &hkey) != ERROR_SUCCESS)
		return 0;

	for(len=16, n = i = 0; RegEnumKeyEx(hkey, i, str, &len, 0, NULL, NULL, &ft) == ERROR_SUCCESS; len = 16, i++, n=0)
	{
		RegOpenKeyEx(hkey, str, NULL, KEY_READ, &hsubkey);
		len = 16;
		if(RegQueryValueEx(hsubkey, REGSZ_LayoutID, NULL, NULL, (LPBYTE) str2, &len) == ERROR_SUCCESS)
		{
			if(strtoul(str2, NULL, 16) == LayoutID) break;  // strtoul - I1546
		}
		RegCloseKey(hsubkey);
	}

	RegCloseKey(hkey);

	n = strtoul(str, NULL, 16); // strtoul - I1546
	return HIWORD(n);
}

WORD HKLToLayoutID(HKL hkl)
{
	if((HIWORD(hkl) & 0xF000) != 0xF000) return 0;

	return HIWORD(hkl) & 0x0FFF;
}
